homeassistant/components/nextcloud/__init__.py

```diff
@@ -27,125 +33,164 @@
 
+    configuration_url: str
 
-# Use recursion to create list of sensors & values based on nextcloud api data
-def get_data_points(api_data, key_path="", leaf=False):
-    """Use Recursion to discover data-points and values.
-
-    Get dictionary of data-points by recursing through dict returned by api until
-    the dictionary value does not contain another dictionary and use the
-    resulting path of dictionary keys and resulting value as the name/value
-    for the data-point.
-
-    returns: dictionary of data-point/values
-    """
-    result = {}
-    for key, value in api_data.items():
-        if isinstance(value, dict):
-            if leaf:
-                key_path = f"{key}_"
-            if not leaf:
-                key_path += f"{key}_"
-            leaf = True
-            result.update(get_data_points(value, key_path, leaf))
-        else:
-            result[f"{DOMAIN}_{key_path}{key}"] = value
-            leaf = False
-    return result
+    def __init__(
+        self, hass: HomeAssistant, ncm: NextcloudMonitor, entry: ConfigEntry
+    ) -> None:
+        """Initialize the Fritzbox Smarthome device coordinator."""
+        self.entry = entry
+        self.ncm = ncm
+        self.url = entry.data[CONF_URL]
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=self.url,
+            update_interval=timedelta(seconds=60),
+        )
+
+    # Use recursion to create list of sensors & values based on nextcloud api data
+    def _get_data_points(
+        self, api_data: dict, key_path: str = "", leaf: bool = False
+    ) -> dict:
+        """Use Recursion to discover data-points and values.
+
+        Get dictionary of data-points by recursing through dict returned by api until
+        the dictionary value does not contain another dictionary and use the
+        resulting path of dictionary keys and resulting value as the name/value
+        for the data-point.
+
+        returns: dictionary of data-point/values
+        """
+        result = {}
+        for key, value in api_data.items():
+            if isinstance(value, dict):
+                if leaf:
+                    key_path = f"{key}_"
+                if not leaf:
+                    key_path += f"{key}_"
+                leaf = True
+                result.update(self._get_data_points(value, key_path, leaf))
+            else:
+                result[f"{DOMAIN}_{key_path}{key}"] = value
+                leaf = False
+        return result
+
+    async def _async_update_data(self) -> dict:
+        """Fetch all Nextcloud data."""
+
+        def _update_data():
+            try:
+                self.ncm.update()
+            except NextcloudMonitorError as ex:
+                raise UpdateFailed from ex
+            return self._get_data_points(self.ncm.data)
+
+        return await self.hass.async_add_executor_job(_update_data)
+
+
+class NextcloudEntity(CoordinatorEntity[NextcloudDataUpdateCoordinator]):
```

Comment:
Can the base entity not be implemented in a follow-up PR?
Config-flow PRs are already quite big.

If it's not possible, then I'd recommend moving this to a separate `entity` module

Reply:
the base entity was a side effect of implementing the update coordinator